pkg/ingester/wal.go

```diff
@@ -156,12 +156,21 @@ func (w *walWrapper) run() {
 		return
 	}
 
-	ticker := time.NewTicker(w.cfg.CheckpointDuration)
-	defer ticker.Stop()
-
+	lastCheckpoint := time.Now().Add(-w.cfg.CheckpointDuration)
 	for {
 		select {
-		case <-ticker.C:
+		case <-w.quit:
+			level.Info(util.Logger).Log("msg", "creating checkpoint before shutdown")
+			if err := w.performCheckpoint(true); err != nil {
+				level.Error(util.Logger).Log("msg", "error checkpointing series during shutdown", "err", err)
+			}
+			return
+		default:
```

Comment:
The problem with this is, we don't run a checkpoint on exit until the sleep + checkpoint completes

Reply:
You are right. I can have the sleep in a `select {}` with `<-time.After()`.

Reply:
(or instead of `default:`, replace it with `<-time.After()`)